### PR TITLE
Expression pushdown optimization

### PR DESCRIPTION
**Pull request by dai-chen**
*Issue #, if available:*

*Description of changes:* 

 1. **New script language**: Register the expression in new query engine as a script language in Elasticsearch. In this case, we don't need to maintain painless source for each expression.
 2. **The first half**: serialize the expression and send it along with DSL request.
 3. **The second half**: the code in DSL is deserialized to run on each node as if built-in painless script.
 4. **Expression visitor**: Add expression node visitor because both first and second half needs to visit expression tree.

*Limitation*:

 1. **Lucene query optimization**: For now, the entire expression would be pushed down to a script query. There will be another PR for Lucene query API pushdown soon.
 2. **Multi-field name**: Assumption is made that the inner field name is always "keyword", so "text.keyword" is used to get doc value for multi-field.
 3. **Serialization**: for now JDK serializer is in use. Due to not all expressions serializable, ex. arithmetic expression, pipelining filter operator will be used as before if serialization error happens.

*Testing*: Add UT and pass all ITs. Because of explain API missing in new engine, new IT will be added later.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/data/value/ElasticsearchExprValueFactory.java [hunk 1]
```diff
@@ -121,6 +124,49 @@ private ExprValue construct(String field, JsonNode value) {
...
+    } else if (type.equals(TIMESTAMP)) {
+      if (value instanceof Number) {
+        return constructTimestamp((Long) value);
+      } else if (value instanceof Instant) {
+        return constructTimestamp((Instant) value);
+      } else {
+        return constructTimestamp(String.valueOf(value));
+      }
+    } else if (ES_TEXT.isCompatible(type)) {
+      return new ElasticsearchExprTextValue((String) value);
```
Comment (penghuo) [inline elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/data/value/ElasticsearchExprValueFactory.java:162]: Add some comments?

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/ElasticsearchIndex.java [hunk 2]
```diff
@@ -87,6 +92,20 @@ public PhysicalPlan implement(LogicalPlan plan) {
      * index scan.
      */
     return plan.accept(new DefaultImplementor<ElasticsearchIndexScan>() {
+          @Override
+          public PhysicalPlan visitFilter(LogicalFilter node, ElasticsearchIndexScan context) {
```
Comment (penghuo) [inline elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/ElasticsearchIndex.java:96]: What happen if there are multiple filter, e.g. source=index | filter a = 1 | filter b = 1

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/ExpressionScriptEngine.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import lombok.RequiredArgsConstructor;
+import org.elasticsearch.script.FilterScript;
+import org.elasticsearch.script.ScriptContext;
+import org.elasticsearch.script.ScriptEngine;
+
+/**
+ * Custom expression script engine that supports using our expression code in DSL
```
Comment (penghuo) [inline elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/ExpressionScriptEngine.java:32]: our expression -> core engine expression?

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                                SearchLookup lookup,
+                                LeafReaderContext context,
+                                Map<String, Object> params) {
+    super(params, lookup, context);
+    this.expression = expression;
+  }
+
+  @Override
+  public boolean execute() {
+    // Check we ourselves are not being called by unprivileged code.
```
Comment (penghuo) [inline elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java:69]: we ourserlves?

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    } else {
+      return value;
+    }
+  }
+
+  private ExprValue evaluateExpression(Map<String, ExprValue> valueEnv) {
+    ExprTupleValue tupleValue = ExprTupleValue.fromExprValueMap(valueEnv);
+    ExprValue result = expression.valueOf(tupleValue.bindingTuples());
+
+    if (result.isNull()) {
```
Comment (penghuo) [inline elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java:157]: Missing handling?

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+  @Override
+  public boolean execute() {
+    // Check we ourselves are not being called by unprivileged code.
+    SpecialPermission.check();
+
+    return AccessController.doPrivileged((PrivilegedAction<Boolean>) () -> {
+      Set<ReferenceExpression> fields = extractFields(expression);
+      ElasticsearchExprValueFactory valueFactory = buildValueFactory(fields);
+      Map<String, ExprValue> valueEnv = buildValueEnv(fields, valueFactory);
```
Comment (penghuo) [inline elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java:75]: NP, Could we impl BindingTuple and encapsulate these logic in that to achive the same concept of valueEnv.

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/data/value/ElasticsearchExprValueFactory.java (hunk 1 above)
Comment (dai-chen) [reply elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/data/value/ElasticsearchExprValueFactory.java:162]: Sure, will add a new `ExprValue` class to differentiate text and multi-field.

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/ElasticsearchIndex.java (hunk 2 above)
Comment (dai-chen) [reply elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/ElasticsearchIndex.java:96]: Good point. Will double check and add IT to make sure this can be handled properly. Thanks!

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java (hunk 5 above)
Comment (dai-chen) [reply elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java:157]: Since no missing concept in JAVA, will add a `MISSING` constant in value factory. And will add test case for `null` field.

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java (hunk 6 above)
Comment (dai-chen) [reply elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java:75]: Sure, will add a new `Environment` class to encapsulate the map data structure. Thanks!

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/ExpressionScriptEngine.java (hunk 3 above)
Comment (dai-chen) [reply elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/ExpressionScriptEngine.java:32]: Done!

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java (hunk 4 above)
Comment (dai-chen) [reply elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java:69]: Done!

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java (hunk 5 above)
Comment (dai-chen) [reply elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java]: I added missing value check here. But I found that there seems no way to differentiate null and missing in doc value. I tried the following index. Both `null_field` and `missing_field` in 2nd doc gives a script doc value object with empty list inside. Will confirm and come back to this later.

```
PUT null_and_missing/_doc/1
{
  "null_field": "abc",
  "missing_field": 456
}

PUT null_and_missing/_doc/2
{
  "null_field": null
}
```

File: elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/ElasticsearchIndex.java (hunk 2 above)
Comment (dai-chen) [reply elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/ElasticsearchIndex.java:96]: Added logic to handle multiple filter operator pushdown and UT/IT to test. We may need to improve our planner soon since current index scan operator assumes all push down is around original data source. I guess if we do `where | stats | where`, the result would be wrong.
